gson/src/test/java/com/google/gson/JsonObjectAsMapSuiteTest.java

```diff
@@ -0,0 +1,87 @@
+package com.google.gson;
+
+import com.google.common.collect.testing.MapTestSuiteBuilder;
+import com.google.common.collect.testing.SampleElements;
+import com.google.common.collect.testing.TestMapGenerator;
+import com.google.common.collect.testing.features.CollectionFeature;
+import com.google.common.collect.testing.features.CollectionSize;
+import com.google.common.collect.testing.features.MapFeature;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import junit.framework.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.AllTests;
+
+/**
+ * Dynamic {@link MapTestSuiteBuilder Map test suite} for {@link JsonObject#asMap()}. This
+ * complements {@link JsonObjectAsMapTest}, which can cover some cases which are not covered here,
+ * e.g. making sure changes in the {@code Map} view are visible in the {@code JsonObject}.
+ */
+@RunWith(AllTests.class)
+public class JsonObjectAsMapSuiteTest {
+  private static class MapGenerator implements TestMapGenerator<String, JsonElement> {
+    @Override
+    public SampleElements<Entry<String, JsonElement>> samples() {
+      return new SampleElements<>(
+          Map.entry("one", JsonNull.INSTANCE),
+          Map.entry("two", new JsonPrimitive(true)),
+          Map.entry("three", new JsonPrimitive("test")),
+          Map.entry("four", new JsonArray()),
+          Map.entry("five", new JsonObject()));
+    }
+
+    @Override
+    public Map<String, JsonElement> create(Object... elements) {
+      JsonObject object = new JsonObject();
+      for (Object element : elements) {
+        @SuppressWarnings("unchecked")
```

Comment:
nit: I would tend to cast to `Entry<?, ?>` and then cast the results of `getKey()` and `getValue()` so that the casts can be clearly checked at runtime, avoiding the need for a suppression. (They probably are checked at runtime anyway below? I'd have to check the bytecode.) I have no doubt that we do it the way you do here in a ton of existing code, though. And you'll always be stuck with suppressions elsewhere because of our use of arrays....